### ISSUE TITLE
cuda manual_seed doesn't fully set Philox RNG

### ISSUE THREAD

**Issue by t-vi** [pytorch/pytorch#12669]
## 🐛 Bug

Manual seed doesn't seem to tie down the Philox RNG, but only the MTGP32 one.

## To Reproduce

```python
import torch
torch.cuda.manual_seed(1234)
l = torch.cuda.get_rng_state()
torch.bernoulli(torch.full((5,5), 0.5, device='cuda'))
torch.cuda.manual_seed(1234)
l2 = torch.cuda.get_rng_state()
(l==l2).all().item()
```
## Expected behavior

Output 1, i.e. manual_seed sets the full RNG state.

## Additional context

As we ported a few random generation things, notably dropout and bernoulli to ATen, seeding doesn't help them much.

This has been reported on the forums: https://discuss.pytorch.org/t/manual-seed-cannot-make-dropout-deterministic-on-cuda-for-pytorch-1-0-preview-version/27281/

Thank you Changmao Cheng for reporting with a small example!
Also thanks to @albanD and @ptrblck for the discussion and verification.

## Solution

I'll produce a patch.

**Comment by syed-ahmed** [comment]
Hi @t-vi . Thanks for helping in the forum and reporting the bug! I am working on a full refactor of the state Generators in PyTorch (https://github.com/syed-ahmed/pytorch/commit/2fd52166f0b7ef23b2dde60b6308342cc5184d8a) and will be submitting it in the next couple of days. I intend to fix this bug in my PR and also clear some of the confusions in documentation and asymmetry in  the CPU and CUDA generator design in the backend.

**Comment by t-vi** [comment]
@syed-ahmed Ah, sorry, I didn't mean to pre-empt your work, it just sounded like a pretty grave bug - given that we move more and more things to ATen. I'm not sure I scanned your code right, but would it be reasonable to reset the `offset` when setting the seed? That's what my patch does.

I look forward to having better random in PyTorch, thank you for your hard work!

**Comment by syed-ahmed** [comment]
No worries! I think your commit is good to go to solve the bug for now. The offset for `curandStatePhilox` wasn't being reset when doing `THCRandom_manualSeed`, which your code takes care off. My change utilizes Runtian's Philox module from the JIT compiler and I would have to ensure that my `manualSeed()` call, resets the offset as well.
